pkg/logs/input/docker/launcher.go

```diff
@@ -55,13 +59,37 @@ func NewLauncher(sources *config.LogSources, services *service.Services, pipelin
 		stop:               make(chan struct{}),
 		erroredContainerID: make(chan string),
 		lock:               &sync.Mutex{},
+		dockerRetrierStop:  make(chan struct{}, 1),
 	}
-	err := launcher.setup()
-	if err != nil {
+
+	var retryStrategy retry.Strategy
+	if shouldRetry {
+		retryStrategy = retry.RetryCount
+	} else {
+		retryStrategy = retry.OneTry
+	}
+	launcher.initRetry.SetupRetrier(&retry.Config{
+		Name:          "docker Launcher setup",
+		AttemptMethod: func() error { return launcher.setup() },
+		Strategy:      retryStrategy,
+		RetryCount:    math.MaxInt32,
+		RetryDelay:    30 * time.Second,
+	})
+
+	if err := launcher.initRetry.TriggerRetry(); err != nil && err.RetryStatus == retry.PermaFail {
 		return nil, err
 	}
-	// Sources and services are added after the setup to avoid creating
-	// a channel that will lock the scheduler in case of setup failure
+	go func() {
```

Comment:
Why do we need this code + passing a lambda `AttemptMethod` ?

Reply:
The `retrier` itself doesn’t take care of calling `AttemptMethod` in the background until it succeeds.
Instead, it just provides the `TriggerRetry()` function that acts as a kind of gate keeper to decide whether it is worth invoking `AttemptMethod` or not.

Hence the need for the go routing for handling the effective retries.